src/kgen.cpp

```diff
@@ -1,113 +1,128 @@
 #include <iostream>
 #include <vector>
 #include <algorithm>
+#include <functional>
+
+namespace kgen {
+
+    template<typename T>
+    class gen {
+    private:
+        class _lookback {
```

Comment:
Maybe use a different naming convention, like `lookback_parent` or `lookback_abstract`?. I don't like that the child class is `lookback::_lookback`.

Reply:
Done, `lb_base`